crates/rb-sys/src/stable_api/ruby_3_3.rs

```diff
@@ -268,4 +279,37 @@ impl StableApiDefinition for Definition {
 
         unsafe { crate::rb_thread_wait_for(time) }
     }
+
+    fn rstruct_define(&self, name: &CStr, members: &[&CStr]) -> VALUE {
+        let mut members: Vec<*const c_char> = members
+            .iter()
+            .map(|m| m.as_ptr() as *const c_char)
+            .collect();
+        members.push(std::ptr::null());
+        unsafe { crate::rb_struct_define(name.as_ptr(), members) }
+    }
+
+    unsafe fn rstruct_get(&self, st: VALUE, idx: c_int) -> VALUE {
+        crate::rb_struct_getmember(st, idx as _)
+    }
+
+    unsafe fn rstruct_set(&self, st: VALUE, idx: c_int, value: VALUE) {
+        crate::rb_struct_aset(st, idx as _, value);
+    }
+
+    // unsafe fn rstruct_len(&self, st: VALUE) -> c_long {
+    //     crate::rb_num2long(crate::rb_struct_size(st))
+    // }
+
+    unsafe fn rstruct_len(&self, st: VALUE) -> c_long {
+        let rbasic = st as *const crate::RBasic;
+        if ((*rbasic).flags & RSTRUCT_EMBED_LEN_MASK as VALUE) != 0 {
+            let mut ret = ((*rbasic).flags & RSTRUCT_EMBED_LEN_MASK as VALUE) as c_long;
+            ret >>= RSTRUCT_EMBED_LEN_SHIFT;
+            ret
+        } else {
+            let rstruct = st as *const crate::RStruct;
```

Comment:
Hey @ianks need you help here!

I implemented `rstruct_len` using  `rb_struct_size` and that works fine but then in order to keep perf inline with macros I implemented it by hand using CRuby's implementation but I am struggling to find a way of tweaking the bindings so I get `RStruct`.

Got any pointers?

I did declare `RStruct` in wrapper.h to make it work but that is not probably how it should be done.

Anyways, let me know!

Reply:
You're doing it right. `RStruct` is hidden, so we'll have to manually "bindgen" it hard coding it's `#[repr(C)]` equivalent in rust, like you have already done.

Reply:
Manually bindgen struct `rstruct_len` and `struct_get` seem to be working fine, replaced the dynamically link ones by handwritten ones. Two remaining.